### ISSUE TITLE
Deserialize any class?

### ISSUE THREAD

**Issue by benlabbeus** [AdrianStrugala/AvroConvert#50]
Is it possible to automatically generate c# class from the avro schema?
Or deserialize to a dynamic type?

I'm having a very difficult time extracting data from a complex schema. I created code to generate the classes, but for some reason it's barfing when trying to bind to "string" fields and class fields. long's seem to be totally fine.

Any help greatly appreciated!

**Comment by AdrianStrugala** [comment]
Hello,

Unfortunately, AvroConvert doesn't provide a possibility to deserialize objects to dynamic type.

Generation of c# classes from avro schema is also not possible, but it looks like a promising feature to be implemented in the future.

Regards,
Adrian

**Comment by blabz2007** [comment]
If you're interested here's code I wrote to generate c# classes - but I'm having trouble getting the data to bind. Maybe you'll have better luck - 
```

void Main()
{
	byte[] avroFileContent = File.ReadAllBytes("log1.avro");
	
	GenerateClass(avroFileContent);	
}



public void GenerateClass(byte[] avroData)
{
	string schema = AvroConvert.GetSchema(avroData);
	JObject json = (JObject)JsonConvert.DeserializeObject(schema);
	
	List<AvroClass> classes = new List<AvroClass>();
	GetRecordClasses(json, classes);
	//classes.Dump();
	
	StringBuilder sb = new StringBuilder();
	foreach (AvroClass ac in classes)
	{
		sb.AppendLine($"public class {ac.ClassName}");
		sb.AppendLine("{");
		foreach (AvroField field in ac.Fields)
		{
			sb.AppendLine($"	public {field.FieldType} {field.Name} {{get; set;}}");
		}
		sb.AppendLine("}");
		sb.AppendLine();
	}
	
	Console.WriteLine(sb.ToString());

}

private void GetRecordClasses(object json, List<AvroClass> classes)
{
	if (json is JObject)
	{
		JObject parent = json as JObject;
		foreach (var prop in parent)
		{
			if (prop.Key == "type" && prop.Value.ToString() == "record")
			{
				AvroClass c = new AvroClass()
				{
					ClassName = parent["name"].ToString()
				};
				classes.Add(c);
				// Get Fields
				foreach (var field in parent["fields"] as JArray)
				{
					if (field is JObject)
					{
						// Get Field type
						string fieldType = string.Empty;
						if (field["type"] is JValue)
						{
							fieldType = field["type"].ToString();
						}
						else if (field["type"] is JObject)
						{	
							JObject typeObj = field["type"] as JObject;
							if (typeObj["type"].ToString() == "record")
							{
								fieldType = typeObj["name"].ToString();
							}
							else if (typeObj["type"].ToString() == "array")
							{
								// If this is an array of a specific class that's being defined in this area of the json
								if (typeObj["items"] is JObject && ((JObject)typeObj["items"])["type"].ToString() == "record")
								{
									fieldType = ((JObject)typeObj["items"])["name"].ToString() + "[]";
								}
								else if (typeObj["items"] is JValue)
								{
									fieldType = ((JValue)typeObj["items"]).ToString() + "[]";
								}
								else
								{
									throw new Exception($"{field.ToString()}");	
								}

							}
							else
							{
								throw new Exception($"Unable to process field type of {typeObj["type"].ToString()}");		
							}
							
						}
						else if (field["type"] is JArray)
						{
							JArray types = (JArray)field["type"];
							bool isNullable = false;
							foreach (var e in types)
							{
								if (e.ToString() == "null")
									isNullable = true;
							}
							
							if (types.Count > 2)
								throw new Exception($"Unable to determine acceptable data type for {field["type"].ToString()}");
								
							// Is the field type an object that's defined in this spot
							JToken arrayFieldType = types.Where(x => x.ToString() != "null").FirstOrDefault();
							if (arrayFieldType is JValue)
							{
								fieldType = arrayFieldType.ToString();
							}
							else if (arrayFieldType is JObject)
							{
								if (arrayFieldType["type"].ToString() == "array" && arrayFieldType["items"] is JObject && ((JObject)arrayFieldType["items"])["type"].ToString() == "record")
								{
									fieldType = ((JObject)arrayFieldType["items"])["name"].ToString();
								}
								else if (arrayFieldType["type"].ToString() == "array" && arrayFieldType["items"] is JValue)
								{
									fieldType = arrayFieldType["items"].ToString();
								}
								else
								{
									throw new Exception($"Unable to create field");
								}
							}
							else
							{
								throw new Exception($"Unable to create array in array");
							}
						}
						else
						{
							throw new Exception($"Unable to process field type of {field["type"].GetType().Name}");
						}
							
						if (fieldType == "boolean")
							fieldType = "bool";
						if (fieldType == "bytes")
							fieldType = "byte[]";
						/*
						if (fieldType == "scala.Tuple9")
							fieldType = "Tuple9";*/
						c.Fields.Add(new AvroField()
						{
							Name = field["name"].ToString(),
							FieldType = fieldType
						});
					}
					else
					{
						throw new Exception($"Field type {field.GetType().Name} not supported");
					}
				}
			}
			if (prop.Value is JObject)
			{
				GetRecordClasses(prop.Value, classes);
			}
			else if (prop.Value is JArray)
			{
				foreach (var arrayItem in prop.Value as JArray)
				{
					if (arrayItem is JObject)
					{
						GetRecordClasses((JObject)arrayItem, classes);
					}
					else if (arrayItem is JValue)
					{
						// This could be any item in an array
					}
					else
					{
						throw new Exception($"Unhandled newtonsoft type {arrayItem.GetType().Name}");
					}
				}
			}
		}
	}
	else if (json is JArray)
	{
		throw new Exception($"Unhandled array on root level");
	}
	else
	{
		throw new Exception($"Unidentified newtonsoft object type {json.GetType().Name}");	
	}
}

public class AvroClass
{
	public string ClassName {get;set;}
	public List<AvroField> Fields {get;set;} = new List<AvroField>();
}

public class AvroField
{
	public string FieldType {get;set;}
	public string Name {get;set;}
}

```

**Comment by AdrianStrugala** [comment]
Thank you very much for sharing the code! I will work on this feature as a next priority.

**Comment by AdrianStrugala** [comment]
Hello,

Your code worked like a charm!
After just a few changes I included it in 3.2.0 version of AvroConvert under the feature 'GenerateModel'.

Please test it and confirm if it's working as expected.

Thanks,
Adrian
